runtime/executor/src/main/java/edu/snu/onyx/runtime/executor/data/blocktransfer/BlockOutputStream.java

```diff
@@ -326,96 +250,49 @@ private void checkWritableCondition() throws IOException {
    */
   private final class ByteBufOutputStream extends OutputStream {
 
-    @Nullable
-    private ByteBuf byteBuf = null;
-
-    private long streamLength = 0;
-
     @Override
     public void write(final int i) {
-      createByteBufIfNeeded();
+      final ByteBuf byteBuf = channel.alloc().ioBuffer(1, 1);
       byteBuf.writeByte(i);
-      flushIfFull();
-      streamLength++;
+      writeDataFrame(byteBuf, false);
     }
 
     @Override
     public void write(final byte[] bytes, final int offset, final int length) {
-      int cursor = offset;
-      int bytesToWrite = length;
-      while (bytesToWrite > 0) {
-        createByteBufIfNeeded();
-        final int toWrite = Math.min(bytesToWrite, byteBuf.writableBytes());
-        byteBuf.writeBytes(bytes, cursor, toWrite);
-        cursor += toWrite;
-        bytesToWrite -= toWrite;
-        flushIfFull();
-      }
-      streamLength += length;
-    }
-
-    /**
-     * Writes a data frame from {@link ByteBuf} and creates a new one.
-     */
-    @Override
-    public void flush() {
-      if (byteBuf != null && byteBuf.readableBytes() > 0) {
-        writeDataFrame(false);
-      }
+      final ByteBuf byteBuf = channel.alloc().ioBuffer(length, length);
+      byteBuf.writeBytes(bytes, offset, length);
+      writeDataFrame(byteBuf, false);
     }
 
     @Override
     public void close() {
       // should send a frame with "isLastFrame" on to indicate the end of the block stream
-      writeDataFrame(true);
-      if (byteBuf != null) {
-        byteBuf.release();
-        byteBuf = null;
-      }
-    }
-
-    /**
-     * Flushes the buffer if the buffer is full.
-     */
-    private void flushIfFull() {
-      if (byteBuf != null && byteBuf.writableBytes() == 0) {
-        writeDataFrame(false);
-      }
+      writeDataFrame(null, true);
     }
 
     /**
      * Writes a data frame from {@link ByteBuf}.
      *
+     * @param byteBuf     {@link ByteBuf} to write.
      * @param isLastFrame whether or not the frame is the last frame
      */
-    private void writeDataFrame(final boolean isLastFrame) {
+    private void writeDataFrame(final ByteBuf byteBuf, final boolean isLastFrame) {
```

Comment:
Do we really need to maintain the method `writeDataFrame` and giving `null` for `byteBuf` if it is a last frame?

Reply:
We can just make separate methods for non-last frame & last frame both.